server.go

```diff
@@ -0,0 +1,273 @@
+	if len(issues.Errors) > 0 {
+		writeJSONOrBust(w, issues)
+		return
+	}
+
+	state, stateErr := stateForDomain(domain)
+	if stateErr != nil {
+		msg := fmt.Sprintf("Internal error: could not get current domain status. (%s)\n", stateErr)
+		http.Error(w, msg, http.StatusInternalServerError)
+	}
+
+	switch state.Status {
+	case StatusUnknown:
+		fallthrough
+	case StatusRejected:
+		fallthrough
+	case StatusRemoved:
+		putErr := putState(DomainState{
+			Name:           domain,
+			Status:         StatusPending,
+			SubmissionDate: time.Now(),
+		})
+		if putErr != nil {
+			issue := hstspreload.Issue{
+				Code:    "internal.server.preload.save_failed",
+				Summary: "Internal error",
+				Message: "Unable to save to the pending list.",
+			}
+			issues = hstspreload.Issues{
+				Errors:   append(issues.Errors, issue),
+				Warnings: issues.Warnings,
+			}
+		}
+	case StatusPending:
+		formattedDate := state.SubmissionDate.Format("Monday, _2 January 2006")
+		issue := hstspreload.Issue{
+			Code:    "server.preload.already_pending",
+			Summary: "Domain has already been submitted",
+			Message: fmt.Sprintf("Domain is already pending. It was submitted on %s.", formattedDate),
+		}
+		issues = hstspreload.Issues{
+			Errors:   issues.Errors,
+			Warnings: append(issues.Warnings, issue),
+		}
+	case StatusPreloaded:
+		issue := hstspreload.Issue{
+			Code:    "server.preload.already_preloaded",
+			Summary: "Domain is already preloaded",
+			Message: "The domain is already preloaded.",
+		}
+		issues = hstspreload.Issues{
+			Errors:   append(issues.Errors, issue),
+			Warnings: issues.Warnings,
+		}
+	default:
+		issue := hstspreload.Issue{
+			Code:    "internal.server.preload.unknown_status",
+			Summary: "Internal error",
+			Message: "Cannot preload; could not find domain status.",
+		}
+		issues = hstspreload.Issues{
+			Errors:   append(issues.Warnings, issue),
+			Warnings: issues.Warnings,
+		}
+	}
+
+	writeJSONOrBust(w, issues)
+}
+
+func pending(w http.ResponseWriter, r *http.Request) {
+	names, err := domainsWithStatus(StatusPending)
+	if err != nil {
+		msg := fmt.Sprintf("Internal error: could not retrieve pending list. (%s)\n", err)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return
+	}
+
+	fmt.Fprintf(w, "[\n")
+	for i, name := range names {
+		comma := ","
+		if i+1 == len(names) {
+			comma = ""
+		}
+
+		fmt.Fprintf(w, `    { "name": "%s", "include_subdomains": true, "mode": "force-https" }%s
+`, name, comma)
+	}
+	fmt.Fprintf(w, "]\n")
+}
+
+func difference(from []chromiumpreload.Domain, take []chromiumpreload.Domain) (diff []chromiumpreload.Domain) {
+	takeSet := make(map[chromiumpreload.Domain]bool)
+	for _, elem := range take {
+		takeSet[elem] = true
+	}
+
+	for _, elem := range from {
+		if !takeSet[elem] {
+			diff = append(diff, elem)
+		}
+	}
+
+	return diff
+}
+
+func update(w http.ResponseWriter, r *http.Request) {
+	// Get preload list.
+	preloadList, listErr := chromiumpreload.GetLatest()
```

Comment:
(This actually won't run in AppEngine.)

In any case, I believe the state machine for submissions and the idempotency of this function guarantee that there will never be race conditions here. And if there are issues, the next update should take care of it.

(Note that this only updates domains that were previously pending, and have now been added to the Chromium code.)